File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

document.addEventListener('turbo:load', function (event) {
  const messages = document.getElementById('messages');

  if (messages) {
    createRoomChannel(messages.dataset.roomId);
    messages.scrollTop = messages.scrollHeight;
  }
})

document.addEventListener('keypress', function (event) {
  if (event.target.matches('#message_body')) {
    const message = event.target.value;
    if (event.keyCode === 13 && message !== '') {
      room.speak(message);
      event.target.value = '';
      event.preventDefault();
    } else if (event.keyCode === 13 && message === '') {
      event.preventDefault();
    }
  }
});


function createRoomChannel(roomId) {
  window.room = consumer.subscriptions.create({channel: "RoomChannel", roomId: roomId}, {
    connected() {
      console.log('Connected to RoomChannel');
    },

    disconnected() {
      console.log('Disconnected from RoomChannel');
    },

    received(data) {
      console.log('Received data from RoomChannel: ' + data['message']);

      const messages = document.getElementById('messages');
      messages.innerHTML += data.message;
      messages.scrollTop = messages.scrollHeight;
    },

    speak: function (message) {
      this.perform('speak', {message: message});
    }
  })
}

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    logger.info "subscribed to RoomChannel, roomID: #{params[:roomId]}"

    @room = Room.find(params[:roomId])
    stream_from "room_channel_#{@room.id}"
    speak('message' => '* * * joined the room * * *')
  end

  def unsubscribed
    logger.info "unsubscribed from RoomChannel"

    speak('message' => '* * * left the room * * *')
  end

  def speak(data)
    logger.info "RoomChannel, speak: #{data.inspect}"

    MessageService.new(body: data['message'], user: current_user, room: @room).perform
  end
end
